Magento_Catalog: gallery - fotorama__caption show/hide

joe-vortex opened the pull request:
### Issue
The fotorama version in 2.2.5 in includes a parameter for `captions` in addition to `showCaption`, which defaults to true and shows the caption below the gallery image, there is no way to control this in the traditional sense, besides hiding with CSS or adding a plugin using the `setOptions()` function.

### Testing
With the blank/luma default view.xml vars:
```
...
    <vars module="Magento_Catalog">
        <var name="gallery">
            <var name="captions">false</var> <!-- Show caption below images -->
        </var>
    </vars>
...
```
The `fotorama__caption` element should be hidden on screen.

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

Commit:
fotorama__caption show/hide

The fotorama version in 2.2.5 in includes a parameter for `captions` in addition to `showCaption`, which defaults to true and shows the caption below the gallery image, there is no way to control this in the traditional sense, besides hiding with CSS or adding a plugin use the `setOptions()` function.

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/not_signed)](https://cla.dev.magento.com/magento/magento2?pullRequest=16941) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=16941) before we can accept your contribution.<br/>

magento-engcom-team commented:
Hi @joe-vortex. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me {$VERSION} instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://github.com/magento/magento2/wiki/Magento-Contributor-Assistant) documentation

Commit:
correcting logic

If captions is true, no need to output it (default),
If captions is false, output `captions: false`

Commit:
adding default var for captions

Commit:
adding default var for captions

miguelbalparda commented:
Thanks for your contribution @joe-vortex! I'll process this one and report back.

magento-engcom-team commented:
Hi @miguelbalparda, thank you for the review.
[ENGCOM-2440](https://jira.corp.magento.com/browse/ENGCOM-2440) has been created to process this Pull Request

magento-engcom-team commented:
@joe-vortex thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

joe-vortex commented:
Thanks @miguelbalparda !

hostep commented:
Heads up guys, this might clash with the following open PR's: https://github.com/magento/magento2/pull/16594 or https://github.com/magento/magento2/pull/15546.

More context: https://github.com/magento/magento2/issues/15009

/cc @gwharton 

Even though this is a different variable `captions` vs `showCaption`, it might be trying to do the same thing.

Just commenting and warning in here, because I was involved in issues regarding the above PR's today which seem to also solve the same thing as this PR is trying to solve.

gwharton commented:
I dont believe we need this captions option. The correct way of not showing the caption below the images is to set showCaption to false in the config. This is in accordance with the Magento developer docs at https://devdocs.magento.com/guides/v2.2/javascript-dev-guide/widgets/widget_gallery.html

Unfortunately there is a bug in the 2.2.5 gallery.phtml which means that it does not handle boolean variables correctly resulting in the showCaption variable never being output, even though it is defined in the default view.xml.

I believe the correct fix for this issue is to apply PR#16594 which not only fixes the gallery.phtml for showCaption, but also for all other booleans in view.xml which arent applied correctly either. 

I have PR#16594 applied to 2.2.5 and do NOT have captions=false in my view.xml and the captions are not displayed under the images.

Note that the fixes in PR #16594 were previously applied back around 2.2.3/2.2.4 time, but were accidentally reverted (by me....... Arghh) which meant that 2.2.5 shipped without them.

gwharton commented:
Also note the new behaviour of the code $block->getVar("name") when the variable contains either the text true or false.

Since 2.2.5

The function will return boolean true if the variable "name" is present and has the text true
The function will return boolean false if the variable "name" is present and has the text false
The function will return boolean false if the variable "name" is not present

There is no way to differentiate between the variable not existing, and it existing but is false.

joe-vortex commented:
@gwharton I noticed this too. The variable default is true and the getvar function would just not output anything. I have tested this by hard setting `showCaption` to false and it's not adding the element in the DOM. The fix you referenced - `"showCaption": <?= /* @escapeNotVerified */ $block->getVar("gallery/caption") ? 'true' : 'false'` would prevent the confusion with not set/false.

If your PR is going to be included in the next release then we could close my PR. This was my first contribution. Better luck next time I guess :frowning_face:

miguelbalparda reviewed (changes requested):
Still discussin between this one and #16594

miguelbalparda commented:
Right now we have 3 different solutions for this issue #16594 #15546 #16941. It seems we will be moving forward with #16594 but we'd like to have some more input if possible.

hostep commented:
> we'd like to have some more input if possible

I'd vote for https://github.com/magento/magento2/pull/16594 because:
- It doesn't introduce a new variable, so documentation doesn't need to be updated
- It was already applied to 2.3-develop in https://github.com/magento/magento2/commit/163d4cc63d19c3a3bd94697dd02c0d96696df799, so it's best to keep 2.2 & 2.3 consistent

miguelbalparda commented:
Closing in favor of #16594. Thanks for your contribution @joe-vortex, sorry for all the back and forth!!!
